torchtitan/parallelisms/parallelize_llama.py

```diff
@@ -129,15 +138,48 @@ def get_tp_parallel_strategy(
     return RowwiseParallel, ColwiseParallel
 
 
+def apply_pipeline_parallelism(model, world_mesh, parallel_dims, job_config: JobConfig):
+    assert (
+        parallel_dims.pp_enabled
+    ), "can't apply pipeline parallelism if it is not enabled"
+
+    if job_config.model.norm_type == "fused_rmsnorm":
+        # TODO(whc) - torch._dynamo.exc.Unsupported: Illegal getattr invocation stride in strict mode
+        # coming from ` if dy.stride(-1) != 1:` in fused_rmsnorm
+        raise NotImplementedError(
+            "fused_rmsnorm not yet compatible with Pipeline Tracer (strides error). Please use layernorm or rmsnorm."
+        )
+    pp_mesh = world_mesh["pp"]
+    stage_idx = pp_mesh.get_local_rank()
+    layers_per_rank = len(model.layers) // parallel_dims.pp
+    split_spec = {
+        f"layers.{i * layers_per_rank}": SplitPoint.BEGINNING
+        for i in range(1, parallel_dims.pp)
+    }
+    # Get example input
+    label_shape = input_shape = (8, 2048)  # TODO
```

Comment:
@kwen2501 any ideas for a clean way to do this in torchtrain? do we expect people to get a batch out of their dataloader and then reset it? or do we expect people to hardcode it?

Reply:
i think what i might do is directly pass input_shape from train.py,

and in train.py i can set input_shape = (job_config.batch_size, job_config.seq_len) or something. is that clean enough?

Reply:
ok pushed a variation on this.

not sure if its better to hide this inside parallelize since we already have job config, or make it explicit from train.py that we are passing input_shape in for some reason

Reply:
Either way sounds okay to me -- eventually, the shape comes the config.